pkg/hub/controllers/cluster/reconciler.go

```diff
@@ -458,3 +489,50 @@ func (r *Reconciler) InjectClient(c client.Client) error {
 	r.Client = c
 	return nil
 }
+
+func (r *Reconciler) handleClusterDeletion(cluster *hubv1alpha1.Cluster, ctx context.Context, req reconcile.Request) (bool, reconcile.Result, error) {
+	log := logger.FromContext(ctx)
+
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) {
+			controllerutil.AddFinalizer(cluster, clusterDeregisterFinalizer)
+			if err := r.Update(ctx, cluster); err != nil {
+				return true, reconcile.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(cluster, clusterDeregisterFinalizer) && !cluster.Status.IsDeregisterInProgress && retryAttempts <= MAX_CLUSTER_DEREGISTRATION_ATTEMPTS {
+			// our finalizer is present, so lets handle any external dependency
+			if err := r.createDeregisterJob(ctx, cluster); err != nil {
+				// unable to deregister the worker operator, return with an error and raise event
+				log.Error(err, "unable to deregister the worker operator")
+				// increment count for retryAttempts
+				retryAttempts++
+				// resetting isDeregisterInProgress to false
+				isDeregisterInProgress := false
```

Comment:
@priyank-upadhyay @YachikaRalhan 
From what i understood from Priyank's explanation is that the IsDeregisterInProgress is used by the controller to maintain some internal state, and it is also being used by the operator to retry the cleanup. But the operator is relying on the retryAttempts var here to check if another attempt needs to be made. The control will reach this point in the code only if the cluster.Status.IsDeregisterInProgress is false (see line 508) and we are again setting it to false and updating the status. So i am not sure why it needs to be set to false by the operator, since its a no-op as the cluster.Status.IsDeregisterInProgress would anyways be false. IMO the controller and operator should not be modifying the same elements in the Status section. The operator should set the hubv1alpha1.RegistrationStatus and the controller should read it and decide what to do next. But we should not have both operator and controller writing to hubv1alpha1.RegistrationStatus. Should be readonly for one and readWrite for the other.